File: source/includes/aggregation/unpack-arrays.js
const { MongoClient } = require("mongodb");

const uri = "<connection string>";
const client = new MongoClient(uri);

async function run() {
  try {
    const aggDB = client.db("agg_tutorials_db");

    // start-coll
    const ordersColl = await aggDB.collection("orders");
    // end-coll

    // start-insert-orders
    await ordersColl.deleteMany({});

    const orderData = [
      {
        order_id: 6363763262239,
        products: [
          {
            prod_id: "abc12345",
            name: "Asus Laptop",
            price: 431,
          },
          {
            prod_id: "def45678",
            name: "Karcher Hose Set",
            price: 22,
          },
        ],
      },
      {
        order_id: 1197372932325,
        products: [
          {
            prod_id: "abc12345",
            name: "Asus Laptop",
            price: 429,
          },
        ],
      },
      {
        order_id: 9812343774839,
        products: [
          {
            prod_id: "pqr88223",
            name: "Morphy Richards Food Mixer",
            price: 431,
          },
          {
            prod_id: "def45678",
            name: "Karcher Hose Set",
            price: 21,
          },
        ],
      },
      {
        order_id: 4433997244387,
        products: [
          {
            prod_id: "def45678",
            name: "Karcher Hose Set",
            price: 23,
          },
          {
            prod_id: "jkl77336",
            name: "Picky Pencil Sharpener",
            price: 1,
          },
          {
            prod_id: "xyz11228",
            name: "Russell Hobbs Chrome Kettle",
            price: 16,
          },
        ],
      },
    ];

    await ordersColl.insertMany(orderData);
    // end-insert-orders

    const pipeline = [];

    // start-unwind
    pipeline.push({
      $unwind: {
        path: "$products",
      },
    });
    // end-unwind

    // start-match
    pipeline.push({
      $match: {
        "products.price": {
          $gt: 15,
        },
      },
    });
    // end-match

    // start-group
    pipeline.push({
      $group: {
        _id: "$products.prod_id",
        product: { $first: "$products.name" },
        total_value: { $sum: "$products.price" },
        quantity: { $sum: 1 },
      },
    });
    // end-group

    // start-set
    pipeline.push({
      $set: {
        product_id: "$_id",
      },
    });
    // end-set

    // start-unset
    pipeline.push({ $unset: ["_id"] });
    // end-unset

    // start-run-agg
    const aggregationResult = await ordersColl.aggregate(pipeline);
    // end-run-agg

    for await (const document of aggregationResult) {
      console.log(document);
    }
  } finally {
    await client.close();
  }
}

run().catch(console.dir);

File: content/manual/v6.0/source/includes/aggregation/aggregation-examples/unpack-arrays/full-files/unpack_arrays.go
package main

import (
	"context"
	"fmt"
	"log"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

// start-structs
type Order struct {
	OrderID  int       `bson:"order_id"`
	Products []Product `bson:"products"`
}

type Product struct {
	ProductID string `bson:"prod_id"`
	Name      string `bson:"name"`
	Price     int    `bson:"price"`
}

// end-structs

func main() {
	const uri = "<connection string>"

	client, err := mongo.Connect(options.Client().ApplyURI(uri))
	if err != nil {
		log.Fatal(err)
	}

	defer func() {
		if err = client.Disconnect(context.TODO()); err != nil {
			log.Fatal(err)
		}
	}()

	aggDB := client.Database("agg_tutorials_db")

	// start-insert-orders
	orders := aggDB.Collection("orders")
	orders.DeleteMany(context.TODO(), bson.D{})

	_, err = orders.InsertMany(context.TODO(), []interface{}{
		Order{
			OrderID: 6363763262239,
			Products: []Product{
				{ProductID: "abc12345", Name: "Asus Laptop", Price: 431},
				{ProductID: "def45678", Name: "Karcher Hose Set", Price: 22},
			},
		},
		Order{
			OrderID: 1197372932325,
			Products: []Product{
				{ProductID: "abc12345", Name: "Asus Laptop", Price: 429},
			},
		},
		Order{
			OrderID: 9812343774839,
			Products: []Product{
				{ProductID: "pqr88223", Name: "Morphy Richards Food Mixer", Price: 431},
				{ProductID: "def45678", Name: "Karcher Hose Set", Price: 21},
			},
		},
		Order{
			OrderID: 4433997244387,
			Products: []Product{
				{ProductID: "def45678", Name: "Karcher Hose Set", Price: 23},
				{ProductID: "jkl77336", Name: "Picky Pencil Sharpene", Price: 1},
				{ProductID: "xyz11228", Name: "Russell Hobbs Chrome Kettle", Price: 16},
			},
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	// end-insert-orders

	// start-unwind
	unwindStage := bson.D{{Key: "$unwind", Value: bson.D{
		{Key: "path", Value: "$products"},
	}}}
	// end-unwind

	// start-match
	matchStage := bson.D{{Key: "$match", Value: bson.D{
		{Key: "products.price", Value: bson.D{{Key: "$gt", Value: 15}}},
	}}}
	// end-match

	// start-group
	groupStage := bson.D{{Key: "$group", Value: bson.D{
		{Key: "_id", Value: "$products.prod_id"},
		{Key: "product", Value: bson.D{{Key: "$first", Value: "$products.name"}}},
		{Key: "total_value", Value: bson.D{{Key: "$sum", Value: "$products.price"}}},
		{Key: "quantity", Value: bson.D{{Key: "$sum", Value: 1}}},
	}}}
	// end-group

	// start-set
	setStage := bson.D{{Key: "$set", Value: bson.D{
		{Key: "product_id", Value: "$_id"},
	}}}
	// end-set

	// start-unset
	unsetStage := bson.D{{Key: "$unset", Value: bson.A{"_id"}}}
	// end-unset

	// start-run-agg
	pipeline := mongo.Pipeline{unwindStage, matchStage, groupStage, setStage, unsetStage}
	cursor, err := orders.Aggregate(context.TODO(), pipeline)
	// end-run-agg

	if err != nil {
		log.Fatal(err)
	}

	defer func() {
		if err := cursor.Close(context.TODO()); err != nil {
			log.Fatalf("failed to close cursor: %v", err)
		}
	}()

	var results []bson.D
	if err = cursor.All(context.TODO(), &results); err != nil {
		log.Fatalf("failed to decode results: %v", err)
	}
	for _, result := range results {
		res, _ := bson.MarshalExtJSON(result, false, false)
		fmt.Println(string(res))
	}
}
